lib/Optimizer/Transforms/LinearCtrlRelations.cpp

```diff
@@ -86,22 +93,28 @@ class ThreadControl : public OpRewritePattern<quake::ToControlOp> {
       }
       const auto coarity = user->getResults().size();
       std::size_t position = 0; // Position of new wire in result tuple.
-      for (Value opnd : user->getOperands()) {
+      constexpr std::size_t Uninitialized = ~0;
+      std::size_t operandNumber = Uninitialized;
+      for (auto iter : llvm::enumerate(user->getOperands())) {
+        Value opnd = iter.value();
         if (isa<quake::WireType>(opnd.getType())) {
           position++;
           continue;
         }
-        if (auto x = opnd.getDefiningOp<quake::ToControlOp>()) {
-          if (x.getOperation() == toCtrl.getOperation())
+        if (auto x = opnd.getDefiningOp<quake::ToControlOp>())
+          if (x.getOperation() == toCtrl.getOperation()) {
+            operandNumber = iter.index();
             break;
-        }
+          }
       }
+      assert(operandNumber != Uninitialized);
```

Comment:
Should we throw a runtime error instead so that non-debug builds throw an error, too?

Reply:
That's a good question. My thought is that if $\alpha$ is a user of $\beta$ but then when we scan $\alpha$'s uses and do not find $\beta$ things have gone off the rails completely. That is, MLIR is itself broken in a fundamental way. That of course is seriously bad and I would expect something like a segfault to happen since pointers are broken.

Which is a long way of saying, I don't know that always checking, even in release mode, buys that much. If this test fails, it is either because something in this pass is wrong or the MLIR framework has completely failed.